csrc/executor_kernel_arg.cpp

```diff
@@ -334,133 +334,140 @@ void KernelArgumentHolder::pushTensorProxy(
       float v32 = (float)v;
       return std::vector<std::byte>(
           (std::byte*)&v32, (std::byte*)&v32 + sizeof(float));
-    } else if (parameter->dtype() == DataType::Half) {
+    } else if (dtype == DataType::Half) {
       at::Half v16 = (at::Half)(float)v;
       return std::vector<std::byte>(
           (std::byte*)&v16, (std::byte*)&v16 + sizeof(at::Half));
-    } else if (parameter->dtype() == DataType::BFloat16) {
+    } else if (dtype == DataType::BFloat16) {
       at::BFloat16 v16 = (at::BFloat16)(float)v;
       return std::vector<std::byte>(
           (std::byte*)&v16, (std::byte*)&v16 + sizeof(at::BFloat16));
     } else {
       TORCH_INTERNAL_ASSERT(
           false,
-          "Tried to run a generated kernel with ",
-          parameter->dtype(),
-          " type, however only float, double, half and bfloat16 are supported.");
+          "Cannot convert double to ",
+          dtype,
+          " type: only half, bfloat16, float and double are supported.");
     }
-  } else if (isComplexType(parameter->dtype())) {
-    std::complex<double> v = pv.as<std::complex<double>>();
-    if (parameter->dtype() == DataType::ComplexDouble) {
+  } else if (argument.is<std::complex<double>>()) {
+    std::complex<double> v = argument.as<std::complex<double>>();
+    if (dtype == DataType::ComplexDouble) {
       return std::vector<std::byte>(
           (std::byte*)&v, (std::byte*)&v + sizeof(std::complex<double>));
-    } else if (parameter->dtype() == DataType::ComplexFloat) {
+    } else if (dtype == DataType::ComplexFloat) {
       std::complex<float> v32 = (std::complex<float>)v;
       return std::vector<std::byte>(
           (std::byte*)&v32, (std::byte*)&v32 + sizeof(std::complex<float>));
     } else {
       TORCH_INTERNAL_ASSERT(
           false,
-          "Tried to run a generated kernel with ",
-          parameter->dtype(),
-          " type, however only complex float and complex double are supported.");
+          "Cannot convert complex double to ",
+          dtype,
+          " type: only complex float and complex double are supported.");
     }
-  } else if (std::holds_alternative<PointerOf>(parameter->dtype().type)) {
-    void* ptr = (void*)pv;
-    return std::vector<std::byte>(
-        (std::byte*)&ptr, (std::byte*)&ptr + sizeof(ptr));
   } else {
     TORCH_INTERNAL_ASSERT(
         false,
-        "Tried to run a generated kernel with unsupported dtype ",
-        parameter->dtype());
+        "Cannot convert ",
+        argument.type().name(),
+        " to kernel argument data.");
+  }
+}
+
+std::vector<std::byte> getKernelArgument(
+    ExpressionEvaluator& ee,
+    Val* parameter,
+    PrimDataType index_type) {
+  TORCH_INTERNAL_ASSERT(parameter != nullptr);
+  PolymorphicValue pv = ee.evaluate(parameter);
+  if (auto tv = dynamic_cast<TensorView*>(parameter)) {
+    if (tv->isCpuScalar()) {
+      return polymorphicValueToBytes(pv, tv->dtype(), index_type);
```

Comment:
This is cool. So if you have a CPU zero-dim tensor you essentially cast it to a real scalar here?

Reply:
It is actually a `CpuScalarTensor<T>`, which is defined as:

```C++
template <typename T>
struct CpuScalarTensor {
  __device__ T& operator[](int i) {
    return data;
  };

  T data;
};
```

and it is essentially just the scalar itself in terms of bytes.